skpro/distributions/qpd.py

```diff
@@ -263,138 +265,78 @@ def __init__(
         self.index = index
         self.columns = columns
 
-        from cyclic_boosting.quantile_matching import J_QPD_S
+        super().__init__(index=index, columns=columns)
 
-        qv_low, qv_median, qv_high = _prep_qpd_params(qv_low, qv_median, qv_high)
+        # precompute parameters for methods
+        phi = _resolve_phi(version)
+        self.phi = phi
 
-        if index is None:
-            index = pd.RangeIndex(qv_low.shape[0])
-            self.index = index
+        qpd_params = _prep_qpd_vars(phi=phi, mode="S", **self._bc_params)
+        self._qpd_params = qpd_params
 
-        if columns is None:
-            columns = pd.RangeIndex(1)
-            self.columns = columns
+    def _ppf(self, p: np.ndarray):
+        """Quantile function = percent point function = inverse cdf."""
+        lower = self._bc_params["lower"]
+        delta = self._qpd_params["delta"]
+        kappa = self._qpd_params["kappa"]
+        c = self._qpd_params["c"]
+        n = self._qpd_params["n"]
+        theta = self._qpd_params["theta"]
 
-        self._shape = (qv_low.shape[0], 1)
+        phi = self.phi
 
-        if version == "normal":
-            self.phi = norm()
-        elif version == "logistic":
-            self.phi = logistic()
-        else:
-            raise Exception("Invalid version.")
-
-        if (np.any(qv_low > qv_median)) or np.any(qv_high < qv_median):
-            warnings.warn(
-                "The SPT values are not monotonically increasing, "
-                "each SPT is sorted by value",
-                stacklevel=2,
-            )
-            idx = np.where((qv_low > qv_median), True, False) + np.where(
-                (qv_high < qv_median), True, False
-            )
-            un_orderd_idx = np.argwhere(idx > 0).tolist()
-            warnings.warn(f"sorted index {un_orderd_idx}", stacklevel=2)
-            for idx in un_orderd_idx:
-                low, mid, high = sorted([qv_low[idx], qv_median[idx], qv_high[idx]])
-                qv_low[idx] = low
-                qv_median[idx] = mid
-                qv_high[idx] = high
-
-        self.qpd = J_QPD_S(
-            alpha=alpha,
-            qv_low=qv_low,
-            qv_median=qv_median,
-            qv_high=qv_high,
-            l=self.lower,
-            version=version,
-        )
-        super().__init__(index=index, columns=columns)
+        in_sinh = np.arcsinh(phi.ppf(p) * delta) + np.arcsinh(n * c * delta)
```

Comment:
In most of the cases, you probably want to do a np.outer rather than an element-wise multiplication here. (In the Cyclic Boosting implementation I made that optional.)

Reply:
not in most cases imo, but I do see how this would be useful in many cases.

This is, however, a framework level concern, so I moved it into a new issue: https://github.com/sktime/skpro/issues/341

Btw, if `np.ndarray` of dim 3 could be passed, the `numpy` native upcasting would already provide the kind of behaviour that is described in #341.